src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/SubstitutedILProvider.cs

```diff
@@ -659,6 +659,67 @@ public MethodIL GetMethodILWithInlinedSubstitutions(MethodIL method)
             return new SubstitutedMethodIL(method.GetMethodILDefinition(), newBody, newEHRegions.ToArray(), debugInfo, newStrings.ToArray());
         }
 
+        private bool TryGetMethodConstantValue(MethodDesc method, out int constant, int level = 0)
+        {
+            method = method.GetTypicalMethodDefinition();
+
+            TypeFlags returnType = method.Signature.ReturnType.Category;
+            if (returnType is < TypeFlags.Boolean or > TypeFlags.UInt32
+                || method.IsIntrinsic
+                || _nestedILProvider.GetMethodIL(method) is not MethodIL methodIL)
+            {
+                constant = 0;
+                return false;
+            }
+
+            var reader = new ILReader(methodIL.GetILBytes());
+            var opcode = reader.ReadILOpcode();
+            switch (opcode)
+            {
+                case ILOpcode.ldc_i4: constant = (int)reader.ReadILUInt32(); break;
+                case ILOpcode.ldc_i4_s: constant = (sbyte)reader.ReadILByte(); break;
+                case >= ILOpcode.ldc_i4_0 and <= ILOpcode.ldc_i4_8: constant = opcode - ILOpcode.ldc_i4_0; break;
+                case ILOpcode.ldc_i4_m1: constant = -1; break;
+
+                case ILOpcode.call:
```

Comment:
Does there need to be any special handling for intrinsics that fold to a constant (like `Vector128.IsHardwareAccelerated` or `Sse.IsSupported`)?

Reply:
I expect those to be trivially rejected since this rejects any methods marked intrinsic:

https://github.com/dotnet/runtime/blob/f8b5e181f1bbe90f2c5a8420b7aaf60522327229/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/SubstitutedILProvider.cs#L667-L673

Reply:
So in that case, helper properties that themselves are just returning constant instrinsics won't be substituted still, right? Is that something we want to support in the future?

In the libraries, we have some properties like:
```csharp
public static bool Vectorizable => (typeof(T) == typeof(double)) || (typeof(T) == typeof(float));
```

and other libraries like `ImageSharp` have APIs like:
```csharp
public static bool IsSupported => AdvSimd.IsSupported;

public sealed override bool IsAvailable => IsSupported;
```

So this comes to mind as something that would be nice to ensure the right substitutions are also happening

Reply:
Yeah, once RyuJIT does the scanning, the scanned view of the program can be much better. RyuJIT already does this substitution when doing codegen; this PR just tries to approximate what RyuJIT will do a bit more so that we avoid some of the damage to whole program view (80-20 principle).

Reply:
👍 Thanks for answering, was mostly just interested in the planned direction here; sounds great!